### PR TITLE
Use name-discarding LLVM context

### PR DESCRIPTION
**Pull request by nagisa**
This is only applicable when neither of --emit=llvm-ir or --emit=llvm-bc are not
requested.

In case either of these outputs are wanted, but the benefits of such context are
desired as well, -Zfewer_names option provides the same functionality regardless
of the outputs requested.

Should be a viable fix for https://github.com/rust-lang/rust/issues/46449

### CODE REVIEW

**Comment by rkruppe** [comment]
Some sort of test would be nice, but the only variant that can reasonably be tested is `--emit llvm-ir -Zfewer-names`, even then a plain codegen test would somewhat fragile. Whatever.

@bors r+

File: src/rustllvm/RustWrapper.cpp [hunk 1]
```diff
@@ -76,11 +76,17 @@ extern "C" char *LLVMRustGetLastError(void) {
...
 
-void LLVMRustSetLastError(const char *Err) {
+extern "C" void LLVMRustSetLastError(const char *Err) {
   free((void *)LastError);
   LastError = strdup(Err);
 }
 
+extern "C" LLVMContextRef LLVMRustContextCreate(bool shouldDiscardNames) {
+  auto ctx = new LLVMContext();
+  ctx->setDiscardValueNames(shouldDiscardNames);
```
Comment (eddyb) [inline src/rustllvm/RustWrapper.cpp]: Why not expose just that method? Seems self-documenting.

Comment (nagisa) [reply src/rustllvm/RustWrapper.cpp]: Just an arbitrary decision. I feel that function with an argument is slightly cleaner for our use-case.

**Comment by michaelwoerister** [comment]
I'll prepare a backport of this soon.

**Comment by michaelwoerister** [comment]
Seems like @alexcrichton beat me to it `:)`

**Pull request merged**